### PR TITLE
fix(step9): cap dreaming array-loop sequence length before scan hang

### PR DESCRIPTION
**Pull request by rama0x1**
## Provider/model disclosure

`--client claude-code --provider anthropic --model claude-sonnet-5`, lane `rama0x1-asi-claude-worker`.

## What's broken

`alberta_framework/steps/step9.py::run_step9_scan` hands its caller-supplied
`rewards`/`next_observations` step arrays straight to `jax.lax.scan` with no
real bound on the leading (step) axis. This is the same "unbounded scan hang"
defect class already fixed this session for other scan-driven array-loop
runners: `core/sarsa.py`, `core/average_reward.py`, and
`core/horde_actor_critic.py`. A hostile or mistaken caller supplying a huge
`rewards`/`next_observations` array forces JAX to trace/compile a scan of that
length, hanging the process well before any step executes. Step 9's per-step
work additionally drives model-based dreaming rollouts (`planning_budget`
candidates per real step), so it is at least as exposed to the hang as its
siblings.

The facade also had no exact-type gate on its inputs: a duck-typed object
whose `.shape`/`.dtype` raised or lied could reach comparisons or the scan
build before its type was confirmed safe.

## Fix

- Adds `_has_step9_trusted_array_type` / `_require_step9_trusted_array`: an
  exact-type gate (`np.ndarray`, `jax.Array`, or `jax.core.Tracer` only,
  checked via `type()` before any attribute access) plus shape/dtype
  validation, matching the established `_trusted_array` convention already
  used in this module family (`alberta_framework/steps/step11.py`,
  `step5.py`).
- Adds a real `_STEP9_SEQUENCE_MAX_STEPS = 10_000` ceiling on `rewards`'
  leading length, matching the ceiling convention established this session in
  `core/sarsa.py::_SARSA_SEQUENCE_MAX_STEPS` and
  `core/horde_actor_critic.py::_HORDE_AC_SEQUENCE_MAX_STEPS` (both `10_000`).
  `config`, `rewards`, and `next_observations` are all validated — type,
  length ceiling, and exact shape/dtype — before `jax.lax.scan` is ever
  constructed.

## Tests (failing-test-first)

Added to `tests/test_step9_hostile_validation.py` (new
`run_step9_scan`-specific section): a `monkeypatch` spy replacing
`alberta_framework.steps.step9.jax.lax.scan` with a function that raises if
called, proving `jax.lax.scan` never runs for:

- a `rewards` array one step over the ceiling (`10_001`);
- a `rewards` array far over the ceiling (`200_000`, the actual hang class);
- a `next_observations` array whose length doesn't match `rewards`;
- a non-exact `config` type;
- a wrong-dtype (`int32`) `rewards` array;
- a hostile duck-typed `rewards` object whose `.shape`/`.dtype` properties
  raise `AssertionError` if touched (proving the type gate runs first, `calls
  == 0`).

Plus a boundary check that a `rewards` array of exactly
`_STEP9_SEQUENCE_MAX_STEPS` still passes the trusted-array gate, and a
control test that a small in-bounds sequence still runs `run_step9_scan`
correctly end to end.

Verified at commit `7e5dafd9a22cbfd3198ecabebe0b8394f54b4ac2` (head of this
branch):

```
uv run pytest tests/test_step9_hostile_validation.py -q -o addopts=""
# 27 passed

uv run pytest tests/test_step9*.py -q -o addopts=""
# 222 passed

uv run ruff check alberta_framework/steps/step9.py tests/test_step9_hostile_validation.py
# All checks passed!

uv run python -m mypy
# step9.py: no errors (2 pre-existing, unrelated errors remain in
# alberta_framework/benchmarks/ipmnist_ceiling.py, not touched by this PR)
```

## Evidence tier

This is a **Fix** (measurement-integrity / hang-guard repair on a scan-driven
array-loop facade), not a benchmark climb — no `outputs/` artifact applies.
Evidence is the failing-then-passing tests plus the verification commands
above.

## Scope

Single file plus its dedicated test file: `alberta_framework/steps/step9.py`
(`_has_step9_trusted_array_type`, `_require_step9_trusted_array`,
`_STEP9_SEQUENCE_MAX_STEPS`, and `run_step9_scan`'s validation) and
`tests/test_step9_hostile_validation.py`. No other behavior changes; no
existing test, threshold, or validator was weakened.

## Collision check

Re-ran `gh pr list --repo elizaOS/asi --state open --json number,title,files
--limit 200` filtered on `step9` immediately before starting work and again
immediately before pushing this branch — no other open PR touches
`alberta_framework/steps/step9.py` either time.

---

### CODE REVIEW

**Comment by lalalune** [comment]
Superseded by merged PR #2060, which consolidates this resource boundary with shared scan/work budgets, combined-work checks, and the broader regression panel.
